### PR TITLE
[WIP] Sse range

### PR DESCRIPTION
**Pull request by ioioioio**
I plugged the [range-sse algorithm](https://github.com/cyb70289/utf8/blob/master/range-sse.c) so we can compare the performances.

The author claims it should not bring any significant improvement on sse. Seems to be true.

Before:
number of iterations 1000 
number of bytes 631515 number of structural chars 55264 ratio 0.088
mem alloc instructions:       1256 cycles:        911 (0.07 %) ins/cycles: 1.38 mis. branches:          5 (cycles/mis.branch 153.85) cache accesses:      30452 (failure          0)
 mem alloc runs at 0.00 cycles per input byte.
stage 1 instructions:    3192517 cycles:     857482 (62.58 %) ins/cycles: 3.72 mis. branches:       1608 (cycles/mis.branch 533.04) cache accesses:      30452 (failure          2)
 stage 1 runs at 1.36 cycles per input byte.
stage 2 instructions:    1577387 cycles:     511731 (37.35 %) ins/cycles: 3.08 mis. branches:       1662  (cycles/mis.branch 307.82)  cache accesses:      55074 (failure          8)
 stage 2 runs at 0.81 cycles per input byte and 9.26 cycles per structural character.
 all stages: 2.17 cycles per input byte.
Estimated average frequency: 4.646 GHz.
Min:  0.000294926 bytes read: 631515 Gigabytes/second: 2.14127

With range-sse:
number of iterations 1000 
number of bytes 631515 number of structural chars 55264 ratio 0.088
mem alloc instructions:       1256 cycles:        885 (0.07 %) ins/cycles: 1.42 mis. branches:          5 (cycles/mis.branch 171.05) cache accesses:      29328 (failure          1)
 mem alloc runs at 0.00 cycles per input byte.
stage 1 instructions:    3094438 cycles:     825148 (61.60 %) ins/cycles: 3.75 mis. branches:       1496 (cycles/mis.branch 551.29) cache accesses:      29328 (failure         62)
 stage 1 runs at 1.31 cycles per input byte.
stage 2 instructions:    1577387 cycles:     513460 (38.33 %) ins/cycles: 3.07 mis. branches:       1767  (cycles/mis.branch 290.56)  cache accesses:      52921 (failure        185)
 stage 2 runs at 0.81 cycles per input byte and 9.29 cycles per structural character.
 all stages: 2.12 cycles per input byte.
Estimated average frequency: 4.636 GHz.
Min:  0.000288942 bytes read: 631515 Gigabytes/second: 2.18561

A neon version might come soon.

### CODE REVIEW

**Comment by lemire** [comment]
At a glance, this looks like progress...
You go from 3192517 instructions down to 3094438 instructions in stage 1... that's a 3% improvement!

Note that this is 3% better over all of stage 1... whereas the new code only affects some of it. That's great stuff if it holds up... gaining 3% would be fantastic.

Which file did you use? twitter.json?

File: include/simdjson/stage1_find_marks_westmere.h [hunk 1]
```diff
@@ -50,12 +49,11 @@ really_inline void check_utf8<Architecture::WESTMERE>(
   if ((_mm_testz_si128(_mm_or_si128(in.v0, in.v1), high_bit)) == 1) {
     // it is ascii, we just check continuation
     state.has_error =
-        _mm_or_si128(_mm_cmpgt_epi8(state.previous.carried_continuations,
+        _mm_or_si128(_mm_cmpgt_epi8(state.previous.first_len,
                                     _mm_setr_epi8(9, 9, 9, 9, 9, 9, 9, 9, 9, 9,
-                                                  9, 9, 9, 9, 9, 1)),
+                                                  9, 9, 9, 3, 2, 1)),
```
Comment (ioioioio) [inline include/simdjson/stage1_find_marks_westmere.h]: That line is wrong.

**Comment by ioioioio** [comment]
yes, I used twitter.json.

File: include/simdjson/simdutf8check_westmere.h [hunk 2]
```diff
@@ -181,9 +176,9 @@ really_inline void check_utf8<Architecture::WESTMERE>(
   if ((_mm_testz_si128(_mm_or_si128(in.v0, in.v1), high_bit)) == 1) {
     // it is ascii, we just check continuation
     state.has_error =
-        _mm_or_si128(_mm_cmpgt_epi8(state.previous.carried_continuations,
+        _mm_or_si128(_mm_cmpgt_epi8(state.previous.first_len,
                                     _mm_setr_epi8(9, 9, 9, 9, 9, 9, 9, 9, 9, 9,
-                                                  9, 9, 9, 9, 9, 1)),
+                                                  9, 9, 9, 3, 2, 1)),
```
Comment (ioioioio) [inline include/simdjson/simdutf8check_westmere.h]: I did some fuzzy tests, found no error, but I'm still not convinced by that line. I found no difference with keeping "9, 9", and I think this is suspicious. I'll have to do more tests before the merge.

Comment (lemire) [reply include/simdjson/simdutf8check_westmere.h]: If only we had a fuzzer tool for such tasks.

Comment (lemire) [reply include/simdjson/simdutf8check_westmere.h]: :-)

Comment (ioioioio) [reply include/simdjson/simdutf8check_westmere.h]: In my understanding of the range algorithm, "9,9,9,9,9,9,9,9,9,9,9,9,9,9,9,1" should not detect any error if the third or the second byte before the end of previous is in the range 0xF0~0xFF and the current 16 bytes are full ascii. Right now I'm trying to be sure we test that case.

I wonder what are the probabilities our fuzzer can generate that case.

Comment (ioioioio) [reply include/simdjson/simdutf8check_westmere.h]: So the line was in fact wrong, but the fuzzer had no difficulty to generate the problematic documents. And I'm a bit ashamed to write down what I was doing wrong.

**Comment by ioioioio** [comment]
The last commit does not change the number of instructions. (who knows)

**Comment by lemire** [comment]
Thank you @ioioioio

**Comment by lemire** [comment]
Made obsolete by https://github.com/lemire/simdjson/pull/365

**Pull request closed without merging**